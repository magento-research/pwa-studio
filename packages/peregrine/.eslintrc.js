const config = {
    parser: 'babel-eslint',
    extends: ['@magento'],
    rules: {
        "no-undef": "off"
    }
};

module.exports = config;
