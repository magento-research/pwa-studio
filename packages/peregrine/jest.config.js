module.exports = {
    displayName: 'Peregrine',
    setupFiles: ['<rootDir>/scripts/shim.js', '<rootDir>/scripts/fetch-mock.js']
};
