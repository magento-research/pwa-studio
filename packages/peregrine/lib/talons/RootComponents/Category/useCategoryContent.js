import { useCallback, useEffect, useState, useRef } from 'react';
import { useLazyQuery, useQuery } from '@apollo/client';

import mergeOperations from '../../../util/shallowMerge';
import { useAppContext } from '../../../context/app';

import DEFAULT_OPERATIONS from './categoryContent.gql';
import { useFilterState } from '../../FilterModal';

const DRAWER_NAME = 'filter';

/**
 * Returns props necessary to render the categoryContent component.
 *
 * @param {object} props.data - The results of a getCategory GraphQL query.
 *
 * @returns {object} result
 * @returns {number} result.categoryId - This category's ID.
 * @returns {string} result.categoryName - This category's name.
 * @returns {object} result.filters - The filters object.
 * @returns {func}   result.handleLoadFilters - A callback function to signal the user's intent to interact with the filters.
 * @returns {func}   result.handleOpenFilters - A callback function that actually opens the filter drawer.
 * @returns {object} result.items - The items in this category.
 * @returns {bool}   result.loadFilters - Whether or not the user has signalled their intent to interact with the filters.
 */
export const useCategoryContent = props => {
    const { categoryId, data, pageSize = 6 } = props;

    const operations = mergeOperations(DEFAULT_OPERATIONS, props.operations);

    const {
        getCategoryContentQuery,
        getProductFiltersByCategoryQuery
    } = operations;

    const placeholderItems = Array.from({ length: pageSize }).fill(null);
    const [loadFilters, setLoadFilters] = useState(false);
    const [showFiltersModal, setShowFiltersModal] = useState(false);
    const [filterState] = useFilterState();
    const [{ drawer }, { toggleDrawer, closeDrawer }] = useAppContext();
    const prevDrawer = useRef(null);

    const handleLoadFilters = useCallback(() => {
        setLoadFilters(true);
    }, [setLoadFilters]);

    const handleOpenFilters = useCallback(() => {
        setShowFiltersModal(true);
        toggleDrawer(DRAWER_NAME);
    }, [setShowFiltersModal, toggleDrawer]);

    const handleCloseFilters = useCallback(() => {
        setShowFiltersModal(false);
        closeDrawer(DRAWER_NAME);
    }, [setShowFiltersModal, closeDrawer]);

    const [getFilters, { data: filterData }] = useLazyQuery(
        getProductFiltersByCategoryQuery,
        {
            fetchPolicy: 'cache-and-network',
            nextFetchPolicy: 'cache-first'
        }
    );

    const { data: categoryData } = useQuery(getCategoryContentQuery, {
        fetchPolicy: 'cache-and-network',
        nextFetchPolicy: 'cache-first',
        skip: !categoryId,
        variables: {
            id: categoryId
        }
    });

    useEffect(() => {
        if (categoryId) {
            getFilters({
                variables: {
                    categoryIdFilter: {
                        eq: categoryId
                    }
                }
            });
        }
    }, [categoryId, getFilters]);

    // Focus for the button that opens filters should be set
    // when filters just applied and filters drawer just closed
    useEffect(() => {
        prevDrawer.current = drawer;
    }, [drawer, filterState]);

    const filters = filterData ? filterData.products.aggregations : null;
    const items = data ? data.products.items : placeholderItems;
    const totalPagesFromData = data
        ? data.products.page_info.total_pages
        : null;
    const categoryName = categoryData ? categoryData.category.name : null;
    const categoryDescription = categoryData
        ? categoryData.category.description
        : null;

    return {
        categoryName,
        categoryDescription,
        filters,
        showFiltersModal,
        handleLoadFilters,
        handleOpenFilters,
        handleCloseFilters,
        items,
        loadFilters,
        totalPagesFromData
    };
};
