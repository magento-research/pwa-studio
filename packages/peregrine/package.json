{
  "name": "@magento/peregrine",
  "version": "3.0.0",
  "publishConfig": {
    "access": "public"
  },
  "description": "The core runtime of Magento PWA",
  "main": "lib/index.js",
  "scripts": {
    "build": " ",
    "build:ci": " ",
    "build:cjs": " ",
    "build:dev": " ",
    "clean": " ",
    "storybook": "start-storybook -p 9001 -c .storybook",
    "storybook:build": "build-storybook -c .storybook -o storybook-dist",
    "watch": " "
  },
  "repository": "github:magento-research/pwa-studio",
  "author": "Magento Commerce",
  "license": "(OSL-3.0 OR AFL-3.0)",
  "bugs": {
    "url": "https://github.com/magento-research/pwa-studio/issues"
  },
  "homepage": "https://github.com/magento-research/pwa-studio/tree/master/packages/peregrine#readme",
  "devDependencies": {
    "@storybook/react": "~4.1.13",
    "intl": "~1.2.5",
    "react": "~16.8.3",
    "react-apollo": "~2.5.3",
    "react-dom": "~16.8.3",
    "react-router-dom": "~5.0.0",
    "react-test-renderer": "~16.8.6",
    "rimraf": "~2.6.3"
  },
  "peerDependencies": {
    "@babel/runtime": "^7.4.2",
    "react": "^16.8.0",
    "react-apollo": "^2.5.3",
    "react-dom": "^16.8.0",
    "react-router-dom": "~5.0.0"
  },
  "engines": {
    "node": "8.10.0 || >=10.7.0",
    "yarn": ">=1.12.0"
  },
  "module": "lib/index.js",
  "jsnext:main": "lib/index.js",
  "es2015": "lib/index.js",
  "sideEffects": false
}
