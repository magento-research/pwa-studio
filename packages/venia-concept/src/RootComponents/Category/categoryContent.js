import React, { Component } from 'react';
import HtmlHead from 'src/components/HtmlHead';
import classify from 'src/classify';
import Gallery from 'src/components/Gallery';
import Pagination from 'src/components/Pagination';
import defaultClasses from './category.css';

class CategoryContent extends Component {
    render() {
        const { classes, pageControl, data, pageSize } = this.props;
        const items = data ? data.category.products.items : null;
        const title = data ? data.category.name : null;
        const description = data ? data.category.description : null;

        return (
            <article className={classes.root}>
                <HtmlHead
                    title={data.category.name}
                    meta_title={data.category.meta_title}
                    meta_keywords={data.category.meta_keywords}
                    meta_description={data.category.meta_description}
                />
                <h1 className={classes.title}>{title}</h1>
                <section>
                    {/* TODO: Switch to RichContent component from Peregrine when merged */}
                    <span
                        dangerouslySetInnerHTML={{
                            __html: description
                        }}
                    />
                </section>
                <section className={classes.gallery}>
                    <Gallery data={items} title={title} pageSize={pageSize} />
                </section>
                <div className={classes.pagination}>
                    <Pagination pageControl={pageControl} />
                </div>
            </article>
        );
    }
}

export default classify(defaultClasses)(CategoryContent);
