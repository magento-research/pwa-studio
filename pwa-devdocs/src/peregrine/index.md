---
title: Peregrine
---

The Peregrine project is a collection of UI Components for Magento PWA projects.
Use, extend, or remix these components to create a unique Magento PWA storefront.

Visit the [`Peregrine`] repository to contribute to this project.

[`Peregrine`]: https://github.com/magento-research/pwa-studio/tree/master/packages/peregrine
