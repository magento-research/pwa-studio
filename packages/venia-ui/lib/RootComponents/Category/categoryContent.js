import React, { Fragment, Suspense } from 'react';
import { shape, string } from 'prop-types';

import { useCategoryContent } from '@magento/peregrine/lib/talons/RootComponents/Category';

import { mergeClasses } from '../../classify';
import { Title } from '../../components/Head';
import Breadcrumbs from '../../components/Breadcrumbs';
import Gallery from '../../components/Gallery';
import Pagination from '../../components/Pagination';
import defaultClasses from './category.css';

const FilterModal = React.lazy(() => import('../../components/FilterModal'));

const CategoryContent = props => {
    const { data, pageControl } = props;

    const talonProps = useCategoryContent({ data });

    const {
        categoryId,
        categoryName,
        filters,
        handleLoadFilters,
        handleOpenFilters,
        items,
        loadFilters,
        pageTitle
    } = talonProps;

    const classes = mergeClasses(defaultClasses, props.classes);

    const header = filters ? (
        <div className={classes.headerButtons}>
            <button
                className={classes.filterButton}
                onClick={handleOpenFilters}
                onFocus={handleLoadFilters}
                onMouseOver={handleLoadFilters}
                type="button"
            >
                {'Filter'}
            </button>
        </div>
    ) : null;

    const modal =
        filters && loadFilters ? <FilterModal filters={filters} /> : null;

    return (
        <Fragment>
            <Breadcrumbs categoryId={categoryId} />
            <Title>{pageTitle}</Title>
            <article className={classes.root}>
                <h1 className={classes.title}>
                    <div className={classes.categoryTitle}>{categoryName}</div>
                </h1>
                {header}
                <section className={classes.gallery}>
                    <Gallery items={items} />
                </section>
                <div className={classes.pagination}>
                    <Pagination pageControl={pageControl} />
                </div>
                <Suspense fallback={null}>{modal}</Suspense>
            </article>
        </Fragment>
    );
};

export default CategoryContent;

CategoryContent.propTypes = {
    classes: shape({
        filterContainer: string,
        gallery: string,
        headerButtons: string,
        pagination: string,
        root: string,
        title: string
    })
};
