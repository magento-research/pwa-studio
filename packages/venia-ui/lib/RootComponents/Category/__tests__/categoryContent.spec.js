import React from 'react';
import { createTestInstance } from '@magento/peregrine';
import { useCategoryContent } from '@magento/peregrine/lib/talons/RootComponents/Category';
import CategoryContent from '../categoryContent';

jest.mock('@magento/peregrine/lib/context/app', () => {
    const state = {};
    const api = {
        toggleDrawer: jest.fn()
    };
    const useAppContext = jest.fn(() => [state, api]);
    return { useAppContext };
});

jest.mock('../../../components/Head', () => ({
    HeadProvider: ({ children }) => <div>{children}</div>,
    StoreTitle: () => 'Title'
}));

jest.mock('@magento/peregrine/lib/talons/RootComponents/Category', () => ({
    useCategoryContent: jest.fn()
}));

jest.mock('../../../components/Breadcrumbs', () => 'Breadcrumbs');
jest.mock('../../../components/Gallery', () => 'Gallery');
jest.mock('../../../components/Pagination', () => 'Pagination');
jest.mock('../../../components/SortedByContainer', () => 'SortedByContainer');
jest.mock('../../../components/FilterModalOpenButton', () => 'FilterModalOpenButton');
jest.mock('../NoProductsFound', () => 'NoProductsFound');

const classes = {
    root: 'a',
    title: 'b',
    gallery: 'c',
    pagination: 'd'
};

const sortProps = [
    { sortDirection: '', sortAttribute: '', sortText: '' },
    jest.fn()
];

const talonProps = {
    categoryName: 'Name',
    categoryDescription: 'test',
    filters: {},
    items: {
        id: 1
    },
    totalPagesFromData: 1
};

test('renders the correct tree', () => {
    const data = {
        products: {
            items: {
                id: 1
            },
            page_info: {
                total_pages: 1
            }
        }
    };
    useCategoryContent.mockReturnValueOnce(talonProps);
    const instance = createTestInstance(
        <CategoryContent
            pageControl={{}}
            data={data}
            pageSize={6}
            sortProps={sortProps}
            classes={classes}
        />
    );

    expect(instance.toJSON()).toMatchSnapshot();
});

test('renders empty page', () => {
    const data = {
        products: {
            items: null,
            page_info: {
                total_pages: 0
            }
        }
    };
    useCategoryContent.mockReturnValueOnce({
        ...talonProps,
        categoryName: 'Empty Name',
        totalPagesFromData: 0
    });
    const instance = createTestInstance(
        <CategoryContent
            pageControl={{}}
            data={data}
            pageSize={6}
            sortProps={sortProps}
            classes={classes}
        />
    );

    expect(instance.toJSON()).toMatchSnapshot();
});
