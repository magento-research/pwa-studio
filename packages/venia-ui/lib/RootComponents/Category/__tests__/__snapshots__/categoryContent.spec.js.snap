// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the correct tree 1`] = `
Array [
  <Breadcrumbs />,
  "Title",
  <article
    className="a"
  >
    <h1
      className="b"
    >
      <div>
        Name
      </div>
    </h1>
    <div>
      
      <div>
        <button
          onClick={[Function]}
        >
          Sort
        </button>
      </div>
    </div>
    <div>
      Items sorted by 
      <span>
        
      </span>
    </div>
    <section
      className="c"
    >
      <Gallery
        items={
          Object {
            "id": 1,
          }
        }
      />
    </section>
    <div
      className="d"
    >
      <Pagination
        pageControl={Object {}}
      />
    </div>
  </article>,
]
`;
